pennylane/ops/op_math/prod.py

```diff
@@ -415,6 +427,177 @@ def _swappable_ops(op1, op2, wire_map: dict = None) -> bool:
+                self.add(prod_factor)
+        elif isinstance(factor, Sum):
+            self._remove_pauli_factors(wires=factor.wires)
+            self._remove_non_pauli_factors(wires=factor.wires)
+            self._factors += (factor.summands,)
+        elif not isinstance(factor, qml.Identity):
+            if isinstance(factor, SProd):
+                self.global_phase *= factor.scalar
+                factor = factor.base
+            wires = factor.wires
+            if isinstance(factor, (qml.Identity, qml.PauliX, qml.PauliY, qml.PauliZ)):
+                self._add_pauli_factor(factor=factor, wires=wires)
+                self._remove_non_pauli_factors(wires=wires)
+            else:
+                self._add_non_pauli_factor(factor=factor, wires=wires)
+                self._remove_pauli_factors(wires=wires)
+
+    def _add_pauli_factor(self, factor: Operator, wires: List[int]):
+        """Adds the given Pauli operator to the temporary ``self._pauli_factors`` dictionary. If
+        there was another Pauli operator acting on the same wire, the two operators are grouped
+        together using the ``self._pauli_mult`` dictionary.
+
+        Args:
+            factor (Operator): Factor to be added.
+            wires (List[int]): Factor wires. This argument is added to avoid calling
+                ``factor.wires`` several times.
+        """
+        wire = wires[0]
+        op2_name = factor.name
+        old_exponent, old_word = self._pauli_factors.get(wire, (1, "Identity"))
+        exponent, new_word = self._pauli_mult[old_word][op2_name]
+        self._pauli_factors[wire] = old_exponent * exponent, new_word
+
+    def _add_non_pauli_factor(self, factor: Operator, wires: List[int]):
+        """Adds the given non-Pauli factor to the temporary ``self._non_pauli_factors`` dictionary.
+        If there alerady exists an identical operator in the dictionary, the two are grouped
+        together.
+
+        If there isn't an identical operator in the dictionary, all non Pauli factors that act on
+        the same wires are removed and added to the ``self._factors`` tuple.
+
+        Args:
+            factor (Operator): Factor to be added.
+            wires (List[int]): Factor wires. This argument is added to avoid calling
+                ``factor.wires`` several times.
+        """
+        if isinstance(factor, Pow):
+            exponent = factor.z
+            factor = factor.base
+        else:
+            exponent = 1
+        op_hash = factor.hash
+        old_hash, old_coeff, old_op = self._non_pauli_factors.get(wires, [None, None, None])
+        if isinstance(old_op, (qml.RX, qml.RY, qml.RZ)) and factor.name == old_op.name:
+            self._non_pauli_factors[wires] = [
+                op_hash,
+                old_coeff,
+                factor.__class__(factor.data[0] + old_op.data[0], wires),
+            ]
+        elif op_hash == old_hash:
+            self._non_pauli_factors[wires][1] += exponent
```

Comment:
The example with `RX` is hard to reproduce because the `Pow` will be simplified in the first step.

A good example would be:
```pycon
>>> qml.prod(qml.s_prod(1.2, qml.CZ([1,0])), qml.s_prod(3.4, qml.ops.Pow(qml.CZ([1,0]), 2.2))).simplify()
4.08*(CZ**3.2(wires=[1, 0]))
```

Reply:
1. The `SProd` will be removed in the `add` method.
2. We are left with `qml.CZ([1, 0])` and `Pow(qml.CZ([1, 0]), 2.2)`.
3. The first op will be added to `self._non_pauli_factors`.
4. When analysing the second op, we will first remove the `Pow` operator and update `exponent=2.2`. Then we will get the current operator acting on wires `(1, 0)`. When comparing the hashes, since they are the same we will update the exponent to `3.2`.

Hopefully this makes it more clear!

Reply:
This is clear now !